mysql/mysqldump.go

```diff
@@ -0,0 +1,575 @@
+	driver "github.com/pingcap/tidb/types/parser_driver"
+)
+
+var valuesRegexp = regexp.MustCompile("\\((.*?)\\)")
+var insertRegexp = regexp.MustCompile("INSERT\\sINTO\\s(.*?)\\sVALUES\\s")
+var MysqlSpatialDataTypes = []string{"geometrycollection", "multipoint", "multilinestring", "multipolygon", "point", "linestring", "polygon", "geometry"}
+
+// ProcessMySQLDump reads mysqldump data from r and does schema or data conversion,
+// depending on whether conv is configured for schema mode or data mode.
+// In schema mode, ProcessMySQLDump incrementally builds a schema (updating conv).
+// In data mode, ProcessMySQLDump uses this schema to convert MySQL data
+// and writes it to Spanner, using the data sink specified in conv.
+func ProcessMySQLDump(conv *internal.Conv, r *internal.Reader) error {
+	for {
+		startLine := r.LineNumber
+		startOffset := r.Offset
+		b, stmts, err := readAndParseChunk(conv, r)
+		if err != nil {
+			return err
+		}
+		for _, stmt := range stmts {
+			isInsert := processStatements(conv, stmt)
+			internal.VerbosePrintf("Parsed SQL command at line=%d/fpos=%d: %d stmts (%d lines, %d bytes) Insert Statement=%v\n", startLine, startOffset, 1, r.LineNumber-startLine, len(b), isInsert)
+			if isInsert {
+				switch ddlstmt := stmt.(type) {
+				case *ast.InsertStmt:
+					processInsertStmt(conv, ddlstmt)
+				}
+			}
+		}
+		if r.EOF {
+			break
+		}
+	}
+	if conv.SchemaMode() {
+		schemaToDDL(conv)
+		conv.AddPrimaryKeys()
+	}
+	return nil
+}
+
+// readAndParseChunk parses a chunk of mysqldump data, returning the bytes read,
+// the parsed AST (nil if nothing read), error and whether we've hit end-of-file.
+func readAndParseChunk(conv *internal.Conv, r *internal.Reader) ([]byte, []ast.StmtNode, error) {
+	var l [][]byte
+	for {
+		b := r.ReadLine()
+		l = append(l, b)
+		// If we see a semicolon or eof, we're likely to have a command, so try to parse it.
+		// Note: we could just parse every iteration, but that would mean more attempts at parsing.
+		if strings.Contains(string(b), ";") || r.EOF {
+			n := 0
+			for i := range l {
+				n += len(l[i])
+			}
+			s := make([]byte, n)
+			n = 0
+			for i := range l {
+				n += copy(s[n:], l[i])
+			}
+			query := string(s)
+			tree, _, err := parser.New().Parse(query, "", "")
+			if err == nil {
+				return s, tree, nil
+			} else {
+				new_tree, err := handleParseError(conv, query, err, l)
+				if err == false {
+					return s, new_tree, nil
+				}
+			}
+			// Likely causes of failing to parse:
+			// a) complex statements with embedded semicolons e.g. 'CREATE FUNCTION'
+			// b) a semicolon embedded in a multi-line comment, or
+			// c) a semicolon embedded a string constant or column/table name.
+			// We deal with this case by reading another line and trying again.
+			conv.Stats.Reparsed++
+		}
+		if r.EOF {
+			return nil, nil, fmt.Errorf("Error parsing last %d line(s) of input", len(l))
+		}
+	}
+}
+
+// processStatements extracts schema information from MYSQL
```

Comment:
Rename to processStatement since this processes single statements, not list of statements?

Reply:
Done